app/org/sagebionetworks/bridge/dynamodb/DynamoUserConsentDao.java

```diff
@@ -26,156 +26,134 @@
+            throw new BridgeServiceException("Consent not found.", HttpStatus.SC_NOT_FOUND);
         }
-        return hasWithdrawn;
+        activeConsent.setWithdrewOn(withdrewOn);
+        mapper.save(activeConsent);
     }
 
     @Override
     public boolean hasConsented(String healthCode, StudyIdentifier studyIdentifier) {
-        return getUserConsent(healthCode, studyIdentifier) != null;
+        checkArgument(isNotBlank(healthCode));
+        checkNotNull(studyIdentifier);
+        
+        return getActiveUserConsent(healthCode, studyIdentifier) != null;
     }
 
     @Override
-    public UserConsent getUserConsent(String healthCode, StudyIdentifier studyIdentifier) {
-        UserConsent consent = getUserConsent3(healthCode, studyIdentifier.getIdentifier());
+    public UserConsent getActiveUserConsent(String healthCode, StudyIdentifier studyIdentifier) {
+        checkArgument(isNotBlank(healthCode));
+        checkNotNull(studyIdentifier);
+        
+        DynamoUserConsent3 hashKey = new DynamoUserConsent3(healthCode, studyIdentifier.getIdentifier());
+
+        Condition condition = new Condition().withComparisonOperator(ComparisonOperator.NULL);
+        DynamoDBQueryExpression<DynamoUserConsent3> query = new DynamoDBQueryExpression<DynamoUserConsent3>()
+            .withQueryFilterEntry("withdrewOn", condition)
+            .withHashKeyValues(hashKey);
+        
+        List<DynamoUserConsent3> results = mapper.query(DynamoUserConsent3.class, query);
+        return results.isEmpty() ? null : results.get(0);
+    }
+    
+    @Override
+    public UserConsent getUserConsent(String healthCode, StudyIdentifier studyIdentifier, long signedOn) {
+        checkArgument(isNotBlank(healthCode));
+        checkNotNull(studyIdentifier);
+        checkArgument(signedOn > 0L);
+        
+        DynamoUserConsent3 hashKey = new DynamoUserConsent3(healthCode, studyIdentifier.getIdentifier());
+        hashKey.setSignedOn(signedOn);
+        
+        DynamoUserConsent3 consent = mapper.load(hashKey);
         if (consent == null) {
-            consent = getUserConsent2(healthCode, studyIdentifier.getIdentifier());
+            throw new BridgeServiceException("Consent not found.", HttpStatus.SC_NOT_FOUND);   
         }
         return consent;
     }
+    
+    @Override
+    public List<UserConsent> getUserConsentHistory(String healthCode, StudyIdentifier studyIdentifier) {
+        checkArgument(isNotBlank(healthCode));
+        checkNotNull(studyIdentifier);
+        
+        DynamoUserConsent3 hashKey = new DynamoUserConsent3(healthCode, studyIdentifier.getIdentifier());
+        
+        DynamoDBQueryExpression<DynamoUserConsent3> query = new DynamoDBQueryExpression<DynamoUserConsent3>()
+            .withHashKeyValues(hashKey);
+
+        return mapper.query(DynamoUserConsent3.class, query).stream().map(consent -> {
+            return (UserConsent)consent;
+        }).collect(Collectors.toList());
```

Comment:
Minor: I believe you can simplify lambda even further and just say consent -> (UserConsent) consent